From abb3ac8f896c8131491b2811adca9eb428fbf971 Mon Sep 17 00:00:00 2001
From: Guilhem Saurel <guilhem.saurel@laas.fr>
Date: Wed, 27 Aug 2025 15:21:30 +0200
Subject: [PATCH 1/6] python-qt: disable vulnerable qtwebengine by default

webengine is optional:

```
extensions/PythonQt_QtAll/PythonQt_QtAll.pro
24:  qtHaveModule(webenginewidgets):CONFIG += PythonQtWebEngineWidgets
```

So disable it by default to fix build
following https://github.com/NixOS/nixpkgs/pull/435067
---
 pkgs/development/libraries/python-qt/default.nix | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/pkgs/development/libraries/python-qt/default.nix b/pkgs/development/libraries/python-qt/default.nix
index 907452702d61df..d9c594b7b5dad2 100644
--- a/pkgs/development/libraries/python-qt/default.nix
+++ b/pkgs/development/libraries/python-qt/default.nix
@@ -7,6 +7,8 @@
   qtwebengine,
   qtxmlpatterns,
   qttools,
+
+  withWebengine ? false, # vulnerable, so disabled by default
 }:
 
 stdenv.mkDerivation (finalAttrs: {
@@ -24,6 +26,8 @@ stdenv.mkDerivation (finalAttrs: {
     qmake
     qttools
     qtxmlpatterns
+  ]
+  ++ lib.optionals withWebengine [
     qtwebengine
   ];
 

From 6781a958a94c5328ca76c21a5627d222d4c1ada0 Mon Sep 17 00:00:00 2001
From: Guilhem Saurel <guilhem.saurel@laas.fr>
Date: Wed, 27 Aug 2025 18:48:01 +0200
Subject: [PATCH 2/6] python-qt: add Qt6 support

---
 .../libraries/python-qt/default.nix           | 78 +++++++++++++++----
 pkgs/top-level/python-packages.nix            |  2 +
 pkgs/top-level/qt5-packages.nix               |  2 +
 pkgs/top-level/qt6-packages.nix               |  4 +
 4 files changed, 69 insertions(+), 17 deletions(-)

diff --git a/pkgs/development/libraries/python-qt/default.nix b/pkgs/development/libraries/python-qt/default.nix
index d9c594b7b5dad2..c98424226a5389 100644
--- a/pkgs/development/libraries/python-qt/default.nix
+++ b/pkgs/development/libraries/python-qt/default.nix
@@ -1,16 +1,28 @@
 {
   lib,
   stdenv,
+
   fetchFromGitHub,
+
   python3,
   qmake,
-  qtwebengine,
-  qtxmlpatterns,
+
+  qt5compat ? null,
+  qtbase,
+  qtdeclarative,
+  qtsvg,
   qttools,
+  qtwebengine,
+  qtxmlpatterns ? null,
 
-  withWebengine ? false, # vulnerable, so disabled by default
+  qt6Support ? false,
 }:
 
+let
+  qtVersion = lib.getVersion qtbase;
+  pyVersion = lib.getVersion python3;
+  versions = "Qt${lib.versions.major qtVersion}-Python${lib.versions.majorMinor pyVersion}";
+in
 stdenv.mkDerivation (finalAttrs: {
   pname = "python-qt";
   version = "3.6.1";
@@ -24,18 +36,50 @@ stdenv.mkDerivation (finalAttrs: {
 
   nativeBuildInputs = [
     qmake
+    qtdeclarative
+    qtsvg
     qttools
-    qtxmlpatterns
   ]
-  ++ lib.optionals withWebengine [
-    qtwebengine
+  ++ lib.optionals qt6Support [
+    qt5compat
+    qtwebengine # optional and vulnerable in Qt5
+  ]
+  ++ lib.optionals (!qt6Support) [
+    qtxmlpatterns
   ];
 
   buildInputs = [ python3 ];
 
+  env.QTDIR = "${qtbase}"; # Used to find qtcoreversion.h
+
+  # generated cpp is available for many Qt5 versions but not Qt6
+  # ref. https://github.com/MeVisLab/pythonqt#binding-generator
+  preConfigure = lib.optionalString qt6Support (
+    let
+      includePaths = lib.concatMapStringsSep ":" (p: "${p}/include") [
+        qtbase
+        qtdeclarative
+        qtsvg
+        qttools
+        qtwebengine
+      ];
+    in
+    ''
+      pushd generator
+      qmake $qmakeFlags CONFIG+=Release generator.pro
+      make -j $NIX_BUILD_CORES
+      ./pythonqt_generator \
+        --include-paths=${includePaths} \
+        --qt-version=${qtVersion} \
+        qtscript_masterinclude.h \
+        build_all.txt
+      popd
+    ''
+  );
+
   qmakeFlags = [
     "PYTHON_DIR=${python3}"
-    "PYTHON_VERSION=3.${python3.sourceVersion.minor}"
+    "PYTHON_VERSION=${lib.versions.majorMinor pyVersion}"
   ];
 
   dontWrapQtApps = true;
@@ -49,20 +93,20 @@ stdenv.mkDerivation (finalAttrs: {
   '';
 
   preFixup = lib.optionalString stdenv.hostPlatform.isDarwin ''
-    install_name_tool -id \
-      $out/lib/libPythonQt-Qt5-Python3.${python3.sourceVersion.minor}.dylib \
-      $out/lib/libPythonQt-Qt5-Python3.${python3.sourceVersion.minor}.dylib
-    install_name_tool -change \
-               libPythonQt-Qt5-Python3.${python3.sourceVersion.minor}.3.dylib \
-      $out/lib/libPythonQt-Qt5-Python3.${python3.sourceVersion.minor}.3.dylib \
-      -id \
-      $out/lib/libPythonQt_QtAll-Qt5-Python3.${python3.sourceVersion.minor}.dylib \
-      $out/lib/libPythonQt_QtAll-Qt5-Python3.${python3.sourceVersion.minor}.dylib
+    install_name_tool \
+      -id $out/lib/libPythonQt-${versions}.dylib \
+          $out/lib/libPythonQt-${versions}.dylib
+
+    install_name_tool \
+      -change      libPythonQt-${versions}.3.dylib \
+          $out/lib/libPythonQt-${versions}.3.dylib \
+      -id $out/lib/libPythonQt_QtAll-${versions}.dylib \
+          $out/lib/libPythonQt_QtAll-${versions}.dylib
   '';
 
   meta = with lib; {
     description = "PythonQt is a dynamic Python binding for the Qt framework. It offers an easy way to embed the Python scripting language into your C++ Qt applications";
-    homepage = "https://pythonqt.sourceforge.net/";
+    homepage = "https://mevislab.github.io/pythonqt/";
     license = licenses.lgpl21;
     platforms = platforms.all;
     maintainers = with maintainers; [ hlolli ];
diff --git a/pkgs/top-level/python-packages.nix b/pkgs/top-level/python-packages.nix
index 82871be7f8dccd..f2df83eaba1207 100644
--- a/pkgs/top-level/python-packages.nix
+++ b/pkgs/top-level/python-packages.nix
@@ -14991,6 +14991,8 @@ self: super: with self; {
 
   python-qt = toPythonModule (pkgs.python-qt.override { python3 = self.python; });
 
+  python-qt-qt6 = toPythonModule (pkgs.qt6Packages.python-qt.override { python3 = self.python; });
+
   python-rabbitair = callPackage ../development/python-modules/python-rabbitair { };
 
   python-rapidjson = callPackage ../development/python-modules/python-rapidjson { };
diff --git a/pkgs/top-level/qt5-packages.nix b/pkgs/top-level/qt5-packages.nix
index d65f6c3da347f6..710970a284eeeb 100644
--- a/pkgs/top-level/qt5-packages.nix
+++ b/pkgs/top-level/qt5-packages.nix
@@ -158,6 +158,8 @@ makeScopeWithSplicing' {
 
         pulseaudio-qt = callPackage ../development/libraries/pulseaudio-qt { };
 
+        python-qt = callPackage ../development/libraries/python-qt { };
+
         qca = callPackage ../development/libraries/qca {
           inherit (libsForQt5) qtbase;
         };
diff --git a/pkgs/top-level/qt6-packages.nix b/pkgs/top-level/qt6-packages.nix
index 7ee501d1f94e7b..3eee81d0392740 100644
--- a/pkgs/top-level/qt6-packages.nix
+++ b/pkgs/top-level/qt6-packages.nix
@@ -134,6 +134,10 @@ makeScopeWithSplicing' {
         suffix = "qt6";
       };
 
+      python-qt = callPackage ../development/libraries/python-qt {
+        qt6Support = true;
+      };
+
       sailfish-access-control-plugin =
         callPackage ../development/libraries/sailfish-access-control-plugin
           { };

From 76ec17f27b0f1bc079c9bceffa7e5e35b20c0470 Mon Sep 17 00:00:00 2001
From: Guilhem Saurel <guilhem.saurel@laas.fr>
Date: Thu, 28 Aug 2025 10:33:48 +0200
Subject: [PATCH 3/6] qt6Packages.pythonqt: mark broken on darwin

---
 pkgs/development/libraries/python-qt/default.nix | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/pkgs/development/libraries/python-qt/default.nix b/pkgs/development/libraries/python-qt/default.nix
index c98424226a5389..73d109dc636e90 100644
--- a/pkgs/development/libraries/python-qt/default.nix
+++ b/pkgs/development/libraries/python-qt/default.nix
@@ -110,5 +110,7 @@ stdenv.mkDerivation (finalAttrs: {
     license = licenses.lgpl21;
     platforms = platforms.all;
     maintainers = with maintainers; [ hlolli ];
+    # ref. https://github.com/MeVisLab/pythonqt/issues/276
+    broken = (stdenv.hostPlatform.isDarwin && qt6Support);
   };
 })

From e5f47a93ef6c3eade194972cffc47af287206ed0 Mon Sep 17 00:00:00 2001
From: Guilhem Saurel <guilhem.saurel@laas.fr>
Date: Fri, 5 Sep 2025 07:39:35 +0200
Subject: [PATCH 4/6] python-qt: clean

Co-authored-by: Sandro <sandro.jaeckel@gmail.com>
---
 pkgs/development/libraries/python-qt/default.nix | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/pkgs/development/libraries/python-qt/default.nix b/pkgs/development/libraries/python-qt/default.nix
index 73d109dc636e90..1aec28bf0acc67 100644
--- a/pkgs/development/libraries/python-qt/default.nix
+++ b/pkgs/development/libraries/python-qt/default.nix
@@ -111,6 +111,6 @@ stdenv.mkDerivation (finalAttrs: {
     platforms = platforms.all;
     maintainers = with maintainers; [ hlolli ];
     # ref. https://github.com/MeVisLab/pythonqt/issues/276
-    broken = (stdenv.hostPlatform.isDarwin && qt6Support);
+    broken = stdenv.hostPlatform.isDarwin && qt6Support;
   };
 })

From 9d9e4488b83bafea626cace548b6f05043b98039 Mon Sep 17 00:00:00 2001
From: Guilhem Saurel <guilhem.saurel@laas.fr>
Date: Sat, 6 Sep 2025 16:29:04 +0200
Subject: [PATCH 5/6] python-qt: fix inputs

---
 pkgs/development/libraries/python-qt/default.nix | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/pkgs/development/libraries/python-qt/default.nix b/pkgs/development/libraries/python-qt/default.nix
index 1aec28bf0acc67..b90d2b91cfeb06 100644
--- a/pkgs/development/libraries/python-qt/default.nix
+++ b/pkgs/development/libraries/python-qt/default.nix
@@ -36,6 +36,10 @@ stdenv.mkDerivation (finalAttrs: {
 
   nativeBuildInputs = [
     qmake
+  ];
+
+  buildInputs = [
+    python3
     qtdeclarative
     qtsvg
     qttools
@@ -48,8 +52,6 @@ stdenv.mkDerivation (finalAttrs: {
     qtxmlpatterns
   ];
 
-  buildInputs = [ python3 ];
-
   env.QTDIR = "${qtbase}"; # Used to find qtcoreversion.h
 
   # generated cpp is available for many Qt5 versions but not Qt6

From 7fdf0e1fe64ee3e6c4a47a2198431a30b5f9fa9e Mon Sep 17 00:00:00 2001
From: Guilhem Saurel <guilhem.saurel@laas.fr>
Date: Mon, 8 Sep 2025 08:20:39 +0200
Subject: [PATCH 6/6] python-qt: adopt

---
 pkgs/development/libraries/python-qt/default.nix | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/pkgs/development/libraries/python-qt/default.nix b/pkgs/development/libraries/python-qt/default.nix
index b90d2b91cfeb06..a0e4142def7232 100644
--- a/pkgs/development/libraries/python-qt/default.nix
+++ b/pkgs/development/libraries/python-qt/default.nix
@@ -111,7 +111,10 @@ stdenv.mkDerivation (finalAttrs: {
     homepage = "https://mevislab.github.io/pythonqt/";
     license = licenses.lgpl21;
     platforms = platforms.all;
-    maintainers = with maintainers; [ hlolli ];
+    maintainers = with maintainers; [
+      hlolli
+      nim65s
+    ];
     # ref. https://github.com/MeVisLab/pythonqt/issues/276
     broken = stdenv.hostPlatform.isDarwin && qt6Support;
   };
